# Importing Libraries

In [1]:
import pandas as pd
import re
import nltk
nltk.download('stopwords')
from nltk.stem.porter import PorterStemmer
from nltk.corpus import stopwords
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import precision_score, balanced_accuracy_score


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


# Loading Data

In [2]:
df0 = pd.read_csv('/content/Corona_NLP_train.csv', encoding  = 'latin-1')[['OriginalTweet','Sentiment']].drop_duplicates().reset_index(drop = True).iloc[:10000,:]
df0.head()

OriginalTweet           Sentiment
0  @MeNyrbie @Phil_Gahan @Chrisitv https://t.co/i...             Neutral
1  advice Talk to your neighbours family to excha...            Positive
2  Coronavirus Australia: Woolworths to give elde...            Positive
3  My food stock is not the only one which is emp...            Positive
4  Me, ready to go at supermarket during the #COV...  Extremely Negative

# Data Preprocessing/Cleaning

In [3]:
corpus = []
for index, row in df0.iterrows():
  tweet = row['OriginalTweet']
  tweet = re.sub('[^a-zA-Z]',' ',tweet)
  tweet = tweet.lower()
  tweet = tweet.split()

  PS = PorterStemmer()
  all_stop_words = stopwords.words('english')
  
  tweet = [PS.stem(word) for word in tweet if not word in set(all_stop_words)]
  tweet = ' '.join(tweet)
  corpus.append(tweet)

# Creating a bag-of-words model

In [4]:
CV = CountVectorizer(max_features = 800)
X = CV.fit_transform(corpus).toarray()
y = df0.iloc[:,-1].values
y = [i.strip().lower() for i in y]

# Splitting data into training and testing data sets

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.15, random_state = 1)

# Fitting a classification model

In [6]:
LR_classifier = LogisticRegression(max_iter = 1000)
LR_classifier.fit(X_train, y_train)
y_pred = LR_classifier.predict(X_test)

# Calculate Accuracy (balanced_accuracy) 

In [7]:
# precision_score(y_test, y_pred, labels = ['positive'], average='weighted')
print(f"Blanced Accuracy for Logistic Regression is {round(100*balanced_accuracy_score(y_test, y_pred),2)}%")

Blanced Accuracy for Logistic Regression is 51.57%


max_features = 100, balanced_accuracy = 0.3814403486954703

max_features = 500, balanced_accuracy = 0.49247538309985706

max_features = 750, balanced_accuracy = 0.5160931728721675

max_features = 1250, balanced_accuracy = 0.5259547247519542


Blanced Accuracy for Logistic Regression is 51.57%